playlist_manager/playlist_manager/main.py

```diff
@@ -0,0 +1,308 @@
+		for video in new_videos:
+			index = self.find_insert_index(videos, self.playlist_state[playlist], video)
+			self.insert_into_playlist(playlist, video.video_id, index)
+
+	def refresh_playlist(self, playlist):
+		"""Check playlist mirror is in a good state, and fetch it if it isn't.
+		We try to do this with only one page of list output, to save on quota.
+		If the total length does not match (or we don't have a copy at all),
+		then we do a full refresh.
+		"""
+		logging.debug("Fetching first page of playlist {}".format(playlist))
+		query = self.api.list_playlist(playlist)
+		# See if we can avoid further page fetches.
+		if playlist not in self.playlist_state:
+			logging.info("Fetching playlist {} because we don't currently have it".format(playlist))
+		elif query.is_complete:
+			logging.debug("First page of {} was entire playlist".format(playlist))
+		elif len(self.playlist_state[playlist]) == query.total_size:
+			logging.debug("Skipping fetching of remainder of playlist {}, size matches".format(playlist))
+			return
+		else:
+			logging.warning("Playlist {} has size mismatch ({} saved vs {} actual), refetching".format(
+				playlist, len(self.playlist_state[playlist]), query.total_size,
+			))
+		# Fetch remaining pages, if any
+		query.fetch_all()
+		# Update saved copy with video ids
+		self.playlist_state[playlist] = [
+			item['snippet']['resourceId'].get('videoId') # api implies it's possible that non-videos are added
+			for item in query.items
+		]
+
+	def find_insert_index(self, videos, playlist, new_video):
+		"""Find the index at which to insert new_video into playlist such that
+		playlist remains sorted (it is assumed to already be sorted).
+		videos should be a mapping from video ids to video info.
+		"""
+		# To try to behave as best we can in the presence of mis-sorted playlists,
+		# we walk through the list linearly. We insert immediately before the first
+		# item that should be after us in sort order.
+		# Note that we treat unknown items (videos we don't know) as being before us
+		# in sort order, so that we always insert after them.
+		for n, video_id in enumerate(playlist):
+			if video_id not in videos:
+				# ignore unknowns
+				continue
+			video = videos[video_id]
+			# if this video is after new video, return this index
+			if new_video.start_time < video.start_time:
+				return n
+		# if we reach here, it means everything on the list was before the new video
+		# therefore insert at end
+		return len(playlist)
+
+	def insert_into_playlist(self, playlist, video_id, index):
+		"""Insert video into given playlist at given index.
+		Makes the API call then also updates our mirrored copy.
+		"""
+		logging.info("Inserting {} at index {} of {}".format(video_id, index, playlist))
+		self.api.insert_into_playlist(playlist, video_id, index)
+		# Update our copy
+		self.playlist_state.setdefault(playlist, []).insert(0, video_id)
```

Comment:
Isn't this always adding the new video to the start of the local copy, not to the index it should be?

Reply:
fixed